forest/src/lib.rs

```diff
@@ -1,12 +1,605 @@
+        }
+    }
+
+    fn siblings(&self, node: Node<D, L>) -> &[Node<D, L>] {
+        let parent = self.get(node).parent;
+        match parent {
+            None => &self.roots,
+            Some(parent) => self.children(parent),
+        }
+    }
+
+    fn siblings_mut(&mut self, node: Node<D, L>) -> &mut Vec<Node<D, L>> {
+        let parent = self.get(node).parent;
+        match parent {
+            None => &mut self.roots,
+            Some(parent) => self.children_mut(parent),
+        }
+    }
+
+    fn sibling_index(&self, node: Node<D, L>) -> usize {
+        self.siblings(node)
+            .iter()
+            .position(|t| *t == node)
+            .expect("Forest - missing child")
+    }
+}
+
+impl<D, L> Node<D, L> {
+    fn new(index: Index) -> Node<D, L> {
+        Node(index, PhantomData)
+    }
+
+    /// Check if this node is still valid (has not been deleted).
+    pub fn is_valid(self, f: &Forest<D, L>) -> bool {
+        f.arena.contains(self.0)
+    }
+
+    /// Get this node's parent. Returns `None` if already at the root.
+    pub fn parent(self, f: &Forest<D, L>) -> Option<Node<D, L>> {
+        f.get(self).parent
+    }
+
+    /// Get the root of the tree containing this node. (This is the same as
+    /// calling `parent()` repeatedly.)
+    pub fn root(self, f: &Forest<D, L>) -> Node<D, L> {
+        let mut root = self;
+        while let Some(parent) = root.parent(f) {
+            root = parent;
+        }
+        root
+    }
+
+    /// Get the data for this node.
+    pub fn data(self, f: &Forest<D, L>) -> &D {
+        &f.get(self).data
+    }
+
+    /// Mutably get the data for this node.
+    pub fn data_mut(self, f: &mut Forest<D, L>) -> &mut D {
+        &mut f.get_mut(self).data
+    }
+
+    /// Return `true` if this is a leaf node (containing `L`), or `false` if it's
+    /// a branch node (containing children).
+    pub fn is_leaf(self, f: &Forest<D, L>) -> bool {
+        matches!(f.get(self).leaf_or_children, Leaf(_))
+    }
+
+    /// Get the leaf data at this leaf node.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this node is not a leaf.
+    pub fn leaf(self, f: &Forest<D, L>) -> &L {
+        f.leaf(self)
+    }
+
+    /// Mutably get the data at this leaf node.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this node is not a leaf.
+    pub fn leaf_mut(self, f: &mut Forest<D, L>) -> &mut L {
+        f.leaf_mut(self)
+    }
+
+    /// Get this node's children.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this is not a branch node.
+    pub fn children(self, f: &Forest<D, L>) -> &[Node<D, L>] {
+        f.children(self)
+    }
+
+    /// This node's siblings, in order, including itself.
+    /// If this is a root node, its siblings are the roots.
+    pub fn siblings(self, f: &Forest<D, L>) -> &[Node<D, L>] {
+        f.siblings(self)
+    }
+
+    /// Determine this node’s index among its siblings.
+    pub fn sibling_index(self, f: &Forest<D, L>) -> usize {
```

Comment:
Public methods should not treat root nodes as siblings of each other. Private methods should, for convenience. This difference should be mentioned in comments on the private siblings methods.

Reply:
Leaving this as is, because it's mildly awkward for `siblings()` to return a singleton slice. Agreed that this shouldn't be super public behavior, but I'm expecting that `ast` (which wraps this) will treat this case differently anyways.